File: test_files/performance-and-scalability.py
"""
Test: Performance & Scalability
Contains anti-patterns: O(n^2) operations, blocking I/O in loops, unbounded recursion, and heavyweight per-request initialization.
"""

# O(n^2) pattern
def duplicate_work(items):
    out = []
    for i in items:
        for j in items:
            if i == j:
                out.append((i, j))
    return out

# Blocking sleep inside loop
import time

def slow_loop(n):
    results = []
    for i in range(n):
        time.sleep(0.01)  # blocks thread, causing scaling problems
        results.append(i)
    return results

# Unbounded recursion
def recurse(n):
    if n <= 0:
        return 0
    return 1 + recurse(n-1)

if __name__ == '__main__':
    print(len(duplicate_work(list(range(100)))))

File: test_files/performance-and-scalability.java
/*
Test: Performance & Scalability (Java)
Contains O(n^2) patterns, blocking sleep inside loops, unbounded recursion, and inefficient string operations.
*/

public class performance_and_scalability {
    public static java.util.List<int[]> duplicateWork(int[] items) {
        java.util.List<int[]> out = new java.util.ArrayList<>();
        for (int i : items) {
            for (int j : items) {
                if (i == j) {
                    out.add(new int[]{i, j});
                }
            }
        }
        return out;
    }

    public static java.util.List<Integer> slowLoop(int n) throws InterruptedException {
        java.util.List<Integer> results = new java.util.ArrayList<>();
        for (int i = 0; i < n; i++) {
            Thread.sleep(10); // blocks thread, causing scaling problems
            results.add(i);
        }
        return results;
    }

    public static int recurse(int n) {
        if (n <= 0) return 0;
        return 1 + recurse(n - 1); // unbounded recursion can cause stack overflow
    }

    public static void main(String[] args) throws Exception {
        System.out.println(duplicateWork(java.util.stream.IntStream.range(0, 100).toArray()).size());
    }
}
